core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java

```diff
@@ -316,6 +316,18 @@ public void writeStringArrayNullable(@Nullable String[] array) throws IOExceptio
         }
     }
 
+    /**
+     * Writes a string array, for nullable string, writes it as 0 (empty string).
+     */
```

Comment:
I'm not a fan of having both of these methods. The difference between them is too subtle. I think we should instead replace `writeStringArrayNullable` with `writeOptionalStringArray` as the former is trappy and could easily lead to bugs. Serialisation code should always serialise and de-serialise to the same thing.